// react-style-tag/lib/constants.js
'use strict';

exports.__esModule = true;
exports.SUPPORTS_BEFORE_UPDATE_SNAPSHOT = exports.REACT_MINOR_VERSION = exports.IS_PRODUCTION = exports.BEAUTIFY_OPTIONS = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @constant {Object} BEAUTIFY_OPTIONS the options to pass to cssbeautify
 */
var BEAUTIFY_OPTIONS = exports.BEAUTIFY_OPTIONS = {
  autosemicolon: true,
  indent: '  '
};

/**
 * @constant {boolean} IS_PRODUCTION is the runtime in the production environment
 */
// external dependencies
var IS_PRODUCTION = exports.IS_PRODUCTION = !!(process && process.env && process.env.NODE_ENV === 'production');

/**
 * @constant {number} REACT_MINOR_VERSION the numeric major.minor version of React
 */
var REACT_MINOR_VERSION = exports.REACT_MINOR_VERSION = +_react2.default.version.split('.').slice(0, 2).join('.');

/**
 * @constant {boolean} SUPPORTS_BEFORE_UPDATE_SNAPSHOT is getSnapshotBeforeUpdate supported by the React version
 */
var SUPPORTS_BEFORE_UPDATE_SNAPSHOT = exports.SUPPORTS_BEFORE_UPDATE_SNAPSHOT = !isNaN(REACT_MINOR_VERSION) && REACT_MINOR_VERSION >= 16.3;
