// react-style-tag/lib/styles.js
'use strict';

exports.__esModule = true;
exports.getRenderedStyles = exports.getProcessedStyles = undefined;

var _cssbeautify = require('cssbeautify');

var _cssbeautify2 = _interopRequireDefault(_cssbeautify);

var _stylis = require('stylis');

var _stylis2 = _interopRequireDefault(_stylis);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @function getProcessedStyles
 *
 * @description
 * get the styles processed by stylis
 *
 * @param {string} rawStyle the style to process
 * @param {Object} props the props passed to the component
 * @param {boolean} props.isCompressed is compressed CSS output requested
 * @param {boolean} props.isPrefixed is vendor-prefixed CSS requested
 * @returns {string} the processed styles
 */
var getProcessedStyles = exports.getProcessedStyles = function getProcessedStyles(rawStyle, _ref) {
  var isCompressed = _ref.isCompressed,
      isPrefixed = _ref.isPrefixed;
  return new _stylis2.default({
    compress: isCompressed,
    global: false,
    keyframe: false,
    prefix: isPrefixed
  })('', rawStyle);
};

/**
 * @function getRenderedStyles
 *
 * @description
 * get the styles rendered in the HTML tag
 *
 * @param {string} rawStyle the style to process
 * @param {Object} props the props passed to the component
 * @param {boolean} props.isMinified is minified CSS output requested
 * @returns {string} the styles to use in the rendered tag
 */


// constants
// external dependencies
var getRenderedStyles = exports.getRenderedStyles = function getRenderedStyles(rawStyle, props) {
  return props.isMinified ? getProcessedStyles(rawStyle, props) : (0, _cssbeautify2.default)(getProcessedStyles(rawStyle, props), _constants.BEAUTIFY_OPTIONS);
};
